src/pgrn-custom-scan.c

```diff
@@ -147,9 +149,43 @@ PGrnCreateCustomScanState(CustomScan *cscan)
 	return (Node *) &(state->parent);
 }
 
+static Relation
+PGrnChooseIndex(Relation table_rel, int errorLevel)
+{
+	// todo: Support pgroonga_condition() index specification.
+	// todo: Implementation of the logic for choosing which index to use.
+	ListCell *lc;
+	List *indexes;
+
+	if (!table_rel)
+		return NULL;
+
+	indexes = RelationGetIndexList(table_rel);
+	foreach (lc, indexes)
+	{
+		Oid indexId = lfirst_oid(lc);
+
+		Relation index = RelationIdGetRelation(indexId);
+		if (!PGrnIndexIsPGroonga(index))
+		{
+			RelationClose(index);
+			continue;
+		}
+		return index;
+	}
+	return NULL;
+}
+
 static void
-PGrnBeginCustomScan(CustomScanState *cscanstate, EState *estate, int eflags)
+PGrnBeginCustomScan(CustomScanState *node, EState *estate, int eflags)
 {
+	PGrnScanState *state = (PGrnScanState *) node;
+	CustomScanState *customScanState = (CustomScanState *) state;
+	Relation index =
+		PGrnChooseIndex(customScanState->ss.ss_currentRelation, ERROR);
+	grn_obj *sourcesTable = PGrnLookupSourcesTable(index, ERROR);
+	elog(LOG, "DEBUG: custom-scan: %s", PGrnInspect(sourcesTable));
+	RelationClose(index);
```

Comment:
I’m not sure whether we need to close the `index` in the `EndCustomScan` step.
According to the PostgreSQL documentation, this index value isn’t stored in the `CustomScanState *node`,
so we don’t need to worry about closing it, do we?
https://www.postgresql.org/docs/current/custom-scan-execution.html#CUSTOM-SCAN-EXECUTION-CALLBACKS

Reply:
This `index` variable does not need to be closed in EndCustomScan().
This is because it is only used in PGrnBeginCustomScan() and closed here.

EndCustomScan() closes the resource set to `CustomScanState` (`PGrnScanState` in this code) as you said.
There is currently no value set for `PGrnScanState`, so EndCustomScan() does not close anything.